cmd/templ/generatecmd/cmd.go

```diff
@@ -135,23 +135,54 @@ func (cmd Generate) Run(ctx context.Context) (err error) {
 	events := make(chan fsnotify.Event)
 	// Count of events currently being processed by the event handler.
 	var eventsWG sync.WaitGroup
-	// Used to check that the event handler has completed.
-	var eventHandlerWG sync.WaitGroup
 	// For errs from the watcher.
 	errs := make(chan error)
 	// Tracks whether errors occurred during the generation process.
 	var errorCount atomic.Int64
-	// For triggering actions after generation has completed.
-	postGeneration := make(chan *GenerationEvent, 256)
 	// Used to check that the post-generation handler has completed.
 	var postGenerationWG sync.WaitGroup
 	var postGenerationEventsWG sync.WaitGroup
 
+	handleCanceled := func(rw *watcher.RecursiveWatcher) {
+		cmd.Log.Debug("Context cancelled, closing watcher")
+		if err := rw.Close(); err != nil {
+			cmd.Log.Error("Failed to close watcher", slog.Any("error", err))
+		}
+		cmd.Log.Debug("Waiting for events to be processed")
+		eventsWG.Wait()
+		cmd.Log.Debug(
+			"All pending events processed, waiting for pending post-generation events to complete",
+		)
+		postGenerationEventsWG.Wait()
+		cmd.Log.Debug(
+			"All post-generation events processed, deleting watch mode text files",
+			slog.Int64("errorCount", errorCount.Load()),
+		)
+		fileEvents := make(chan fsnotify.Event)
+		go func() {
+			if err := watcher.WalkFiles(ctx, cmd.Args.Path, cmd.WatchPattern, fileEvents); err != nil {
+				cmd.Log.Error("Post dev mode WalkFiles failed", slog.Any("error", err))
+				errs <- FatalError{Err: fmt.Errorf("failed to walk files: %w", err)}
+				return
+			}
+			close(fileEvents) // TODO should this be defer'red?
```

Comment:
Should this close be in a defer, since an error could cause an early return and never close the channel?